File: uniffi_bindgen/src/bindings/kotlin/templates/SetTemplate.kt
{%- let inner_type_name = inner_type|type_name(ci) %}

/**
 * @suppress
 */
public object {{ ffi_converter_name }}: FfiConverterRustBuffer<Set<{{ inner_type_name }}>> {
    override fun read(buf: ByteBuffer): Set<{{ inner_type_name }}> {
        val len = buf.getInt()
        return buildSet<{{ inner_type_name }}>(len) {
            repeat(len) {
                add({{ inner_type|read_fn }}(buf))
            }
        }
    }

    override fun allocationSize(value: Set<{{ inner_type_name }}>): ULong {
        val spaceForLength = 4UL
        val spaceForItems = value.map { {{ inner_type|allocation_size_fn }}(it) }.sum()
        return spaceForLength + spaceForItems
    }

    override fun write(value: Set<{{ inner_type_name }}>, buf: ByteBuffer) {
        buf.putInt(value.size)
        value.iterator().forEach {
            {{ inner_type|write_fn }}(it, buf)
        }
    }
}

File: uniffi_bindgen/src/bindings/swift/templates/SetTemplate.swift
#if swift(>=5.8)
@_documentation(visibility: private)
#endif
fileprivate struct {{ ffi_converter_name }}: FfiConverterRustBuffer {
    public static func write(_ value: {{ type_name }}, into buf: inout [UInt8]) {
        let len = Int32(value.count)
        writeInt(&buf, len)
        for item in value {
            {{ inner_type|write_fn }}(item, into: &buf)
        }
    }

    public static func read(from buf: inout (data: Data, offset: Data.Index)) throws -> {{ type_name }} {
        let len: Int32 = try readInt(&buf)
        var set = {{ type_name }}()
        set.reserveCapacity(Int(len))
        for _ in 0..<len {
            let item = try {{ inner_type|read_fn }}(from: &buf)
            set.insert(item)
        }
        return set
    }
}
